Fix mercury package dependencies and use tarballs

soumagne opened the pull request:
Clean up mercury package and add boostsys, shared variants
Use tarballs instead of git tags
Fix openpa dependency on gcc < 4.9
Fix boost dependency

soumagne commented:
@ax3l @adamjstewart I see that you've been handling the changes wrt the mercury package in past. I'm the main developer of the mercury library, could you please review the changes that I have for that spack package? There are currently a few dependency issues with that package that I fixed in my PR and listed above, the most important one being the dependency on openpa, which is only required when using a compiler that does not support stdatomic, i.e. gcc < 4.9. Also I would prefer if we could use release tarballs instead of git tags. Thanks!

soumagne commented:
Phil @carns, could you also please have a look in case I missed something? Thanks.

chuckatkins commented on var/spack/repos/builtin/packages/mercury/package.py:
```diff
@@ -38,78 +38,63 @@ class Mercury(CMakePackage):
 
         if self.run_tests:
-            args.append('-DBUILD_TESTING=ON')
+            args.extend('-DBUILD_TESTING=ON')
```
Shouldn't these be `cmake_args.append('-DFOO=BAR')`?  If I'm reading the diff correctly the variable name changed from `args` to `cmake_args` and `append` takes a single value while `extend` expects a list.

chuckatkins replied:
Also it seems like these could use the `variant_bool` helper function you defined above.

soumagne replied:
ok thanks I fixed that one.

chuckatkins commented on var/spack/repos/builtin/packages/mercury/package.py:
```diff
@@ -38,78 +38,63 @@ class Mercury(CMakePackage):
-        args = [
-            '-DBUILD_SHARED_LIBS=ON',
+
+        def variant_bool(feature, on='ON', off='OFF'):
```
Nothing at all wrong with this, just a suggestion to make it a little more concise and "pythonic" with a simple lambda:
```python
variant_bool = lambda feature : str(feature in self.spec)
```
Since CMake will happily accept `'True'` and `'False'` as `-DFOO:BOOL=` values.

soumagne replied:
thanks for the suggestion! I will make that change

soumagne commented:
@chuckatkins I have addressed all of your comments in the commit that I pushed. Please let me know if you see anything else that I should fix or that I missed. Thanks!

adamjstewart commented on var/spack/repos/builtin/packages/mercury/package.py:
```diff
@@ -9,21 +9,21 @@ class Mercury(CMakePackage):
+    url = 'https://github.com/mercury-hpc/mercury/releases/download/v1.0.1/mercury-1.0.1.tar.bz2'
     git = 'https://github.com/mercury-hpc/mercury.git'
 
     version('develop', branch='master', submodules=True)
```
Now that #1983 has finally been merged, you can actually rename this version to `master` if you want and it will still sort correctly.

soumagne replied:
Hmm ok I'm not sure what this implies, should I use a URL that points to the master archive tarball?

adamjstewart replied:
Nope. You can keep it as a download from git. Previously, Spack sorted `master` as being older than all other releases, but with #1983, `master` and `develop` are both treated as the newest versions. So you can do:
```suggestion
    version('master', branch='master', submodules=True)
```
and it should just work.

soumagne replied:
oh I see thanks

adamjstewart commented on var/spack/repos/builtin/packages/mercury/package.py:
```diff
@@ -9,21 +9,21 @@ class Mercury(CMakePackage):
-    variant('cci', default=False, description='Use CCI for network transport')
-    variant('mpi', default=False, description='Use MPI for network transport')
-    variant('ofi', default=True,  description='Use libfabric plugin')
+    version('1.0.1', preferred=True, sha256='02febd56c401ef7afa250caf28d012b37dee842bfde7ee16fcd2f741b9cf25b3')
```
`preferred=True` isn't needed here, as Spack already defaults to the newest stable release.

soumagne replied:
ok thanks, will fix that.

adamjstewart commented on var/spack/repos/builtin/packages/mercury/package.py:
```diff
@@ -9,21 +9,21 @@ class Mercury(CMakePackage):
     variant('sm',  default=False, description='Use shared-memory plugin')
-    variant('opa', default=False, description='Use OpenPA for atomics')
-    variant('boost', default=True, description='Use BOOST preprocessor macros')
+    variant('boostsys', default=False, description='Use system-installed Boost libraries')
```
We generally avoid linking to system dependencies, as they result in unreproducible builds. Do you think it would be okay to remove this variant and always depend on boost?

soumagne replied:
I understand but it is not using the "system" boost in that case. Mercury has two options: either it provides and installs the boost preprocessor headers that are included with mercury (so that there is no need to install boost) or in the case where boost is already installed, it won't install the boost prepocessor headers and just use the ones from the boost dependency. So when using the `+boostsys` variant, it will use/install the boost dependency and when not using that variant there won't be any dependency on boost and the headers included with mercury will be installed along with the mercury headers.

adamjstewart replied:
Ah, I see. In that case, this is probably fine, although you may want to rename the description to make it clear that you are using a Spack-installed copy and not a copy installed on the system. Still, there might be some benefit to always depending on boost as you will always have an up-to-date copy of the headers, although I understand that it will increase installation times. I don't personally use `mercury` (I was simply updating the package for @mlawsonca) so I'll defer to your opinion on this one.

soumagne replied:
OK :) I could simply default that variant to true then so that we always depend on boost by default and that'll give a chance to someone who does not want to install boost to use the headers that we provide. I will also update the description to make sure it's clear. Thanks.

adamjstewart commented on var/spack/repos/builtin/packages/mercury/package.py:
```diff
@@ -38,78 +38,54 @@ class Mercury(CMakePackage):
         if '+ofi' in spec:
-            args.append('-DNA_USE_OFI=ON')
+            cmake_args.extend([
+                '-DNA_OFI_GNI_USE_UDREG=%s' % variant_bool('+udreg'),
```
I prefer to use `append()` instead of `extend([])` if you're only adding a single element to a list.

soumagne replied:
sure I will make that change.

soumagne commented:
@adamjstewart sure will do, please feel free to ping me if you have any issues or questions. I'm also using spack myself so it's also in my interest to contribute back :)

carns commented:
Looks great to me.  Perhaps a separate issue, but can we make the sm variant be true by default?  That plugin implemented internally in Mercury and could just be compiled in all cases, right?

soumagne commented:
@carns I addressed your comment in the updated commit that I just pushed to set sm to true by default. Also agree that since there's no external dependency there's no reason not to build it by default.
@adamjstewart I've also addressed all your previous comments in that commit.

adamjstewart commented:
Spoke to @mlawsonca and she said she'll get back to this in ~1.5 weeks. Going to merge for now. Thanks for offering to maintain your own package in Spack!

soumagne commented:
@adamjstewart sounds good, thanks! @mlawsonca please also feel free to create an issue on https://github.com/mercury-hpc/mercury/issues or just let me know if you still have problems.
